core/src/context.rs

```diff
@@ -279,33 +283,31 @@ impl Context {
             .ok_or::<HolochainError>("No top chain header".into())?;
 
         // Get address of first Token Grant entry (return early if none)
-        let addr = state
+        let grants = state
             .agent()
             .chain_store()
-            .iter_type(&Some(top), &EntryType::CapTokenGrant)
-            .next()
-            .ok_or::<HolochainError>("No CapTokenGrant entry type in chain".into())?
-            .entry_address()
-            .to_owned();
+            .iter_type(&Some(top), &EntryType::CapTokenGrant);
 
         // Get CAS
         let cas = state.agent().chain_store().content_storage();
 
-        // Get according Token Grant entry from CAS
-        let entry = get_entry_from_cas(&cas, &addr)?
-            .ok_or::<HolochainError>("Can't get CapTokenGrant entry from CAS".into())?;
-
-        // Make sure entry is a public grant and return it
-        if let Entry::CapTokenGrant(grant) = entry {
-            match grant.cap_type() {
-                CapabilityType::Public => Ok(addr),
-                _ => Err(HolochainError::ErrorGeneric(
-                    "Got CapTokenGrant, but it was not public!".to_string(),
-                )),
+        for grant in grants {
+            let addr = grant.entry_address().to_owned();
+            let entry = get_entry_from_cas(&cas, &addr)?
+                .ok_or::<HolochainError>("Can't get CapTokenGrant entry from CAS".into())?;
+            // if entry is the public grant return it
+            if let Entry::CapTokenGrant(grant) = entry {
+                if grant.cap_type() == CapabilityType::Public
+                    && grant.id() == ReservedCapabilityId::Public.as_str()
+                {
+                    return Ok(addr);
```

Comment:
Just an added comment but we could probably use a fold() to make it look more idiomatic without the early return but no qualms about the way it is done here